app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/AddressController.scala

```diff
@@ -44,30 +46,71 @@ class AddressController @Inject()(
+                                   addressDisplay: views.html.components.address_display
 )(implicit viewConfig: ViewConfig, ec: ExecutionContext)
     extends FrontendController(cc)
     with Logging
     with WithSubscriptionDetailsActions
     with SessionUpdates {
 
+  def enterAddress(): Action[AnyContent] =
+    authenticatedActionWithSubscriptionReady.async { implicit request =>
+    lazy val showEnterAddressPage =
+      request.subscriptionReady.subscriptionDetails.address match {
+        case u: UkAddress =>
+          Ok(enterAddressPage(Address.ukAddressForm.fill(u), routes.SubscriptionController.checkYourDetails()))
+        case _: NonUkAddress =>
+          logger.warn("Expected UK addresses but found non-UK address")
+          errorHandler.errorResult()
+      }
+      if(request.sessionData.addressLookupResult.nonEmpty) {
+        updateSession(sessionStore, request)(_.copy(addressLookupResult = None)).map {
+          case Left(e) =>
+            logger.warn(s"Could not clear addressLookupResult", e)
+            errorHandler.errorResult()
+          case Right(_) =>
+            showEnterAddressPage
+        }
+      } else {
+        showEnterAddressPage
+      }
+
+    }
+
+  def enterAddressSubmit(): Action[AnyContent] =
+    authenticatedActionWithSubscriptionReady.async { implicit request =>
+      Address.ukAddressForm
+        .bindFromRequest()
+        .fold[Future[Result]](
+          formWithErrors => BadRequest(enterAddressPage(formWithErrors, routes.SubscriptionController.checkYourDetails())),
+          storeAddress
+        )
+    }
+
   def enterPostcode(): Action[AnyContent] =
     authenticatedActionWithSubscriptionReady { implicit request =>
-      val form = request.sessionData.addressLookupResult.fold(Postcode.form)(r => Postcode.form.fill(r.postcode))
-      Ok(enterPostcodePage(form, routes.SubscriptionController.checkYourDetails))
+      val form = request.sessionData.addressLookupResult
+        .fold(AddressLookupRequest.form)(r => AddressLookupRequest.form.fill(AddressLookupRequest(r.postcode, r.filter)))
+      Ok(enterPostcodePage(form, routes.SubscriptionController.checkYourDetails()))
     }
 
   def enterPostcodeSubmit(): Action[AnyContent] =
     authenticatedActionWithSubscriptionReady.async { implicit request =>
-      Postcode.form
+      AddressLookupRequest.form
         .bindFromRequest()
         .fold(
-          formWithErrors => BadRequest(enterPostcodePage(formWithErrors, routes.SubscriptionController.checkYourDetails)), { postcode =>
-            if (request.sessionData.addressLookupResult.map(_.postcode).contains(postcode)) {
+          formWithErrors => BadRequest(enterPostcodePage(formWithErrors, routes.SubscriptionController.checkYourDetails())),
+          { case AddressLookupRequest(postcode, filter) =>
+            // TODO: need to search session data with filter also
+            if (request.sessionData.addressLookupResult.map(_.postcode).contains(postcode)
```

Comment:
remove this TODO as you are doing the lookup

Reply:
done